R/abstract-categories.R

```diff
@@ -0,0 +1,224 @@
+
+setNames <- function (x, value) {
+    if (is.null(value) || !is.character(value)) {
+        halt('Names must be of class "character"')
+    }
+    if (!identical(length(x), length(value))) {
+        halt("Invalid names: supplied ", length(value), " names for ",
+             length(x), " categories")
+    }
+    if (any(is.na(value))) {
+        halt("Category names must be non-missing")
+    }
+    x[] <- mapply(setName, x, value=value, SIMPLIFY=FALSE)
+    return(x)
+}
+
+#' @rdname Categories
+#' @export
+setMethod("names", "AbsCats", function (x) {
+    n <- vapply(x, name, character(1))
+    return(n)
+})
+
+#' @rdname Categories
+#' @export
+setMethod("names<-", "AbsCats", setNames)
+
+#' @rdname Categories
+#' @export
+setMethod("[", c("AbsCats", "ANY"), function (x, i, ...) {
+    x@.Data <- x@.Data[i]
+    return(x)
+})
+
+#' @rdname Categories
+#' @export
+setMethod("[", c("AbsCats", "character"), function (x, i, ...) {
+    indices <- match(i, names(x))
+    if (any(is.na(indices))) {
+        halt("subscript out of bounds: ", serialPaste(i[is.na(indices)]))
+    }
+    callNextMethod(x, i=indices)
+})
+
+#' @rdname Categories
+#' @export
+setMethod("[", c("AbsCats", "numeric"), function (x, i, ...) {
+    invalid.indices <- setdiff(abs(i), seq_along(x@.Data))
+    if (length(invalid.indices)) {
+        halt("subscript out of bounds: ", serialPaste(invalid.indices))
+    }
+    x@.Data <- x@.Data[i]
+    return(x)
+})
+
+#' @rdname Categories
+#' @export
+setMethod("[<-", c("AbsCats", "character"), function (x, i, ..., value) {
+    indices <- match(i, names(x))
+    if (any(is.na(indices))) {
+        # if there are no matches, add it on to the end
+        indices <- i
+    }
+    x@.Data[indices] <- value
+    return(x)
+})
+
+#' @rdname Categories
+#' @export
+setMethod("[[", c("AbsCats", "character"), function (x, i, ...) {
+    indices <- match(i, names(x))
+    if (any(is.na(indices))) {
+        halt("subscript out of bounds: ", serialPaste(i[is.na(indices)]))
+    }
+    callNextMethod(x, i=indices)
+})
+
+#' @rdname Categories
+#' @export
+setMethod("[[<-", c("AbsCats", "character"), function (x, i, ..., value) {
+    indices <- match(i, names(x))
+    if (any(is.na(indices))) {
+        # if there are no matches, add it on to the end
+        x@.Data[[i]] <- value
+        return(x)
+    }
+    callNextMethod(x, i=indices, value)
+})
+
+# a version of modifyList that doesn't recurse into the absCats themselves
+modifyCats <- function (x, val) {
+    stopifnot(is.AbsCats(x), is.AbsCats(val))
+    xnames <- names(x)
+    vnames <- names(val)
+    vnames <- vnames[nzchar(vnames)]
+    for (v in vnames) {
+        x[[v]] <- val[[v]]
+    }
+
+    return(x)
+}
+
+############################################
+## Categories specific methods
+############################################
+
+#' @rdname Categories
+#' @export
+setMethod("ids", "AbsCats", function (x) vapply(x, id, integer(1)))
```

Comment:
If these methods are "Categories specific", why are they set on "AbsCats"?

Reply:
I've changed these to be a bit more strict, now they are only defined for the parent class if they are needed. I was defining them for AbsCats, even if they only currently applied to one subclass because I could see a possible future where they would be needed, but we can always upgrade them in the future.

I've also included a vignette describing abstract categories including tables showing what methods are and aren't implemented.